### PR TITLE
Material

### PR DESCRIPTION
**Pull request by houguohua**

### CODE REVIEW

File: cocos/3d/C3DEffect.cpp [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+void C3DEffect::bind()
+{
+	GL_ASSERT( glUseProgram(_program) );
+
+	__currentEffect = this;
+}
+
+C3DEffect* C3DEffect::getCurrentEffect()
+{
+	return __currentEffect;
```
Comment (minggo) [inline cocos/3d/C3DEffect.cpp:486]: __currentEffect -> _currentEffect

File: cocos/3d/C3DEffect.cpp [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+}
+
+C3DEffect* Uniform::getEffect() const
+{
+	return _effect;
+}
+
+const char* Uniform::getName() const
+{
+	return _name.c_str();
```
Comment (minggo) [inline cocos/3d/C3DEffect.cpp:511]: Why not just return std::string?

File: cocos/3d/C3DEffect.cpp [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+	}
+
+	SAFE_DELETE_ARRAY(vshSource);
+	SAFE_DELETE_ARRAY(fshSource);
+
+	return true;
+}
+
+
+bool C3DEffect::load(const char* vshSource, const char* fshSource, const char* defines)
```
Comment (minggo) [inline cocos/3d/C3DEffect.cpp:156]: How about break this function into more small functions?

File: cocos/3d/C3DEffect.h [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+	void bindValue(Uniform* uniform, const C3DMatrix& value);
+
+	/**
+	* Sets a matrix array uniform value.
+	*
+	* @param uniform The uniform to set.
+	* @param values The array to set.
+	* @param count The number of elements in the array.
+	*/
+	void bindValue(Uniform* uniform, const C3DMatrix* values, unsigned int count = 1);
```
Comment (minggo) [inline cocos/3d/C3DEffect.h:123]: I think `GLProgram` also has these functions. Can reuse them?

File: cocos/3d/C3DEffect.cpp (hunk 1 above)
Comment (houguohua) [inline cocos/3d/C3DEffect.cpp:486]: this is global var.

File: cocos/3d/C3DEffect.cpp (hunk 3 above)
Comment (houguohua) [inline cocos/3d/C3DEffect.cpp:156]: good idea.split in into vs,fs

File: cocos/3d/C3DEffect.cpp (hunk 2 above)
Comment (houguohua) [inline cocos/3d/C3DEffect.cpp:511]: ok

File: cocos/3d/C3DEffect.h (hunk 4 above)
Comment (houguohua) [inline cocos/3d/C3DEffect.h:123]: program is one member of effect

Comment (minggo) [inline cocos/3d/C3DEffect.h:123]: Yep, can use `GLProgram` inside. Then don't have re-write the codes for loading vs and fs.

Comment (houguohua) [inline cocos/3d/C3DEffect.h:123]: GLProgram can not satisfy our need.

File: cocos/3d/C3DEffect.cpp (hunk 1 above)
Comment (lvlonggame) [inline cocos/3d/C3DEffect.cpp:486]: all global var need prefix "__"  ?

Comment (minggo) [inline cocos/3d/C3DEffect.cpp:486]: About global variable, i suggest to use `g_xxx`. But it is ok to use `__xxx`.

File: cocos/3d/C3DEffect.cpp (hunk 2 above)
Comment (lvlonggame) [inline cocos/3d/C3DEffect.cpp:511]: I will change this function.

**Pull request merged**